src/wallet/silentpayments.cpp

```diff
@@ -0,0 +1,241 @@
+CPubKey ComputeECDHSharedSecret(const CKey& scan_key, const CPubKey& sender_public_key, const uint256& outpoints_hash)
+{
+    auto tweaked_scan_seckey{scan_key};
+    tweaked_scan_seckey.TweakMultiply(outpoints_hash.begin());
+    CPubKey result = tweaked_scan_seckey.UnhashedECDH(sender_public_key);
+    assert(result.IsValid());
+    return result;
+}
+
+uint256 ComputeTweak(const CPubKey& ecdh_pubkey, const uint32_t output_index)
+{
+    HashWriter h;
+    h.write(Span{ecdh_pubkey});
+    unsigned char num[4];
+    WriteBE32(num, output_index);
+    h << num;
+    return h.GetSHA256();
+}
+
+std::vector<uint256> GetTxOutputTweaks(const CPubKey& spend_pubkey, const CPubKey& ecdh_pubkey, std::vector<XOnlyPubKey> output_pub_keys)
+{
+    // Because a sender can create multiple outputs for us, we first check the outputs vector for an output with
+    // output index 0. If we find it, we remove it from the vector and then iterate over the vector again looking for
+    // an output with index 1, and so on until one of the following happens:
+    //
+    //     1. We have determined all outputs belong to us (the vector is empty)
+    //     2. We have passed over the vector and found no outputs belonging to us
+    //
+
+    bool removed;
+    uint32_t output_index{0};
+    std::vector<uint256> tweaks;
+    do {
+        // We haven't removed anything yet on this pass and if we don't remove anything, we didn't find
+        // any silent payment outputs and should stop checking
+        removed = false;
+        uint256 tweak = ComputeTweak(ecdh_pubkey, output_index);
+        CPubKey computed_sp_pubkey{spend_pubkey};
+        computed_sp_pubkey.TweakAdd(tweak.data());
+        const XOnlyPubKey& sp_pubkey = XOnlyPubKey{computed_sp_pubkey};
+        //const CKey& silent_payment_priv_key = tweakResult.first;
+        output_pub_keys.erase(std::remove_if(output_pub_keys.begin(), output_pub_keys.end(), [&](auto output_pubkey) {
+            if (sp_pubkey == output_pubkey) {
+                // Since we found an output, we need to increment the output index and check the vector again
+                tweaks.emplace_back(tweak);
+                removed = true;
+                output_index++;
+                // Return true so that this output pubkey is removed the from vector and not checked again
+                return true;
+            }
+            return false;
+        }), output_pub_keys.end());
+    } while (!output_pub_keys.empty() && removed);
+    return tweaks;
+}
+
+std::vector<uint256> GetTxOutputTweaks(const CPubKey& spend_pubkey, const CPubKey& ecdh_pubkey, std::vector<XOnlyPubKey> output_pub_keys, const std::pair<uint256, CPubKey> change_label)
+{
+    // Scan first for any outputs to us, and then check if there is any change
+    std::vector<uint256> tweaks = GetTxOutputTweaks(spend_pubkey, ecdh_pubkey, output_pub_keys);
+    std::vector<uint256> labeled_tweaks = GetTxOutputTweaks(change_label.second, ecdh_pubkey, output_pub_keys);
```

Comment:
In 93ff4ce117d4b85128dd55630d2dc9f8efbfb473 "Implement BIP352: Silent Payments"

In the current receiving PR, `change_label` is a pair of the change label and it's corresponding curve point. However this is expecting it as the pair of the change label and the resulting spend pubkey.